lib/concurrent/at_exit.rb

```diff
@@ -0,0 +1,92 @@
+require 'concurrent/logging'
+require 'concurrent/synchronization'
+
+module Concurrent
+
+  # Provides ability to add and remove hooks to be run at `Kernel#at_exit`, order is undefined.
+  # Each hook is executed at most once.
+  class AtExitImplementation < Synchronization::Object
```

Comment:
@pitr-ch I wasn't expecting you to revisit the thread pool registry, but I'm glad you did. I *really* like this. Given the nature of this library, providing synchronization protection for our at_exit handlers makes a lot of sense. I like that this is now a general utility that we can use anywhere in the library should the need arise. This is much more versatile. Thank you!

One small request: Can we please change "hook" to "handler"? I know it's a little pedantic but I think that's a better term.